aya/src/programs/perf_event.rs

```diff
@@ -0,0 +1,115 @@
+use crate::{generated::bpf_prog_type::BPF_PROG_TYPE_PERF_EVENT, sys::perf_event_open};
+
+pub use crate::generated::perf_type_id;
+
+use super::{load_program, perf_attach, LinkRef, ProgramData, ProgramError};
+
+#[derive(Debug, Clone)]
+pub enum SamplePolicy {
+    Period(u64),
+    Frequency(u64),
+}
+
+#[derive(Debug, Clone)]
+#[allow(clippy::enum_variant_names)]
+pub enum PerfEventScope {
+    CallingProcessAnyCpu,
+    CallingProcessOneCpu { cpu: u32 },
+    OneProcessAnyCpu { pid: u32 },
+    OneProcessOneCpu { cpu: u32, pid: u32 },
+    AllProcessesOneCpu { cpu: u32 },
+}
+
+/// A program that can be attached at a perf event.
+///
+/// TODO: Explain the different types of perf events and how to get a list.
+/// Maybe just link to the man page of `perf list`.
+/// (But it's not clear to me how to translate those strings into numbers.)
+///
+/// # Minimum kernel version
+///
+/// TODO: minimum kernel version?
+///
+/// # Examples
+///
+/// ```no_run
+/// # #[derive(Debug, thiserror::Error)]
+/// # enum Error {
+/// #     #[error(transparent)]
+/// #     IO(#[from] std::io::Error),
+/// #     #[error(transparent)]
+/// #     Map(#[from] aya::maps::MapError),
+/// #     #[error(transparent)]
+/// #     Program(#[from] aya::programs::ProgramError),
+/// #     #[error(transparent)]
+/// #     Bpf(#[from] aya::BpfError)
+/// # }
+/// # let mut bpf = aya::Bpf::load(&[], None)?;
+/// use std::convert::TryInto;
+/// use aya::programs::{PerfEvent, PerfEventScope, SamplePolicy };
+/// use aya::util::online_cpus;
+///
+/// let prog: &mut PerfEvent = bpf.program_mut("observe_cpu_clock")?.try_into()?;
+/// prog.load()?;
+///
+/// for cpu in online_cpus()? {
+///     prog.attach(
+///         1, /* PERF_TYPE_SOFTWARE */
+///         0, /* PERF_COUNT_SW_CPU_CLOCK */
```

Comment:
We should generate bindings for ids and flags and whatever else can be passed as config